### ISSUE TITLE
Type could not be determined because it refers to itself

### ISSUE THREAD

**Issue by augustelalande** [microsoft/pyright#9139]
I'm not sure I can describe what's happening better than just sharing the code.

MRE:
```python
from enum import StrEnum


class Foo(StrEnum):
    A = "A"


for _ in range(2):
    baz: dict[Foo, int] = {}

    if Foo.A in baz:
        ...

    for _ in baz.values():
        ...

```

Output:
```
➜ pyright --version
pyright 1.1.383

➜ pyright .\bug.py
c:\Users\Auguste\workspace\test\bug\bug.py
  c:\Users\Auguste\workspace\test\bug\bug.py:11:17 - error: Type of "baz" could not be determined because it refers to itself (reportGeneralTypeIssues)
  c:\Users\Auguste\workspace\test\bug\bug.py:14:14 - error: Type of "baz" could not be determined because it refers to itself (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

**Comment by erictraut** [comment]
Thanks, I'm able to repro the problem and will investigate further.
